agent/taskresource/ssmsecret/ssmsecret.go

```diff
@@ -0,0 +1,476 @@
+	if secret.appliedStatus == resourcestatus.ResourceStatus(SSMSecretStatusNone) {
+		return
+	}
+
+	// Check if the resource transition has already finished
+	if secret.appliedStatus <= knownStatus {
+		secret.appliedStatus = resourcestatus.ResourceStatus(SSMSecretStatusNone)
+	}
+}
+
+// SetAppliedStatus sets the applied status of resource and returns whether
+// the resource is already in a transition
+func (secret *SSMSecretResource) SetAppliedStatus(status resourcestatus.ResourceStatus) bool {
+	secret.lock.Lock()
+	defer secret.lock.Unlock()
+
+	if secret.appliedStatus != resourcestatus.ResourceStatus(SSMSecretStatusNone) {
+		// return false to indicate the set operation failed
+		return false
+	}
+
+	secret.appliedStatus = status
+	return true
+}
+
+// GetKnownStatus safely returns the currently known status of the task
+func (secret *SSMSecretResource) GetKnownStatus() resourcestatus.ResourceStatus {
+	secret.lock.RLock()
+	defer secret.lock.RUnlock()
+
+	return secret.knownStatusUnsafe
+}
+
+// StatusString returns the string of the cgroup resource status
+func (secret *SSMSecretResource) StatusString(status resourcestatus.ResourceStatus) string {
+	return SSMSecretStatus(status).String()
+}
+
+// SetCreatedAt sets the timestamp for resource's creation time
+func (secret *SSMSecretResource) SetCreatedAt(createdAt time.Time) {
+	if createdAt.IsZero() {
+		return
+	}
+	secret.lock.Lock()
+	defer secret.lock.Unlock()
+
+	secret.createdAt = createdAt
+}
+
+// GetCreatedAt sets the timestamp for resource's creation time
+func (secret *SSMSecretResource) GetCreatedAt() time.Time {
+	secret.lock.RLock()
+	defer secret.lock.RUnlock()
+
+	return secret.createdAt
+}
+
+// Create fetches secret value from SSM
+func (secret *SSMSecretResource) Create() error {
+
+	//To fail fast, check execution role first
+	executionCredentials, ok := secret.credentialsManager.GetTaskCredentials(secret.GetExecutionCredentialsID())
+	if !ok {
+		// No need to log here. managedTask.applyResourceState already does that
+		return errors.New("ssm secret resource: unable to find execution role credentials")
+	}
+	iamCredentials := executionCredentials.GetIAMRoleCredentials()
+
+	var wg sync.WaitGroup
+	chanLen := secret.getGoRoutineTotalNum()
+	errorEvents := make(chan error, chanLen)
+	seelog.Infof("ssm secret resource: retrieving secrets for containers in task: [%s]", secret.taskARN)
+	if secret.secretData == nil {
+		secret.secretData = make(map[string]string)
+	}
+
+	for region, secrets := range secret.GetRequiredSecrets() {
+		wg.Add(1)
+		go secret.retrieveSSMSecretValuesByRegion(region, secrets, iamCredentials, &wg, errorEvents)
+	}
+
+	wg.Wait()
+	//get the first error returned and set as terminal reason
+	select {
+	case err := <-errorEvents:
+		secret.setTerminalReason(err.Error())
+		return err
+	default:
+		break
+	}
+	return nil
+}
+
+func (secret *SSMSecretResource) getGoRoutineTotalNum() int {
```

Comment:
so this is for the total number of regions? should this be the same as the number of keys in your region:secrets map?

Reply:
No. This should be the same number as goroutines we spin up to call GetParameters API. So it may not match the number of keys since one region can have multiple goroutines. This is the max number since we dedupe the secrets based on valueFrom + region.